HPCC-17855 DynamicESDL support for get_data_from

mayx opened the pull request:
- Output DataFor "verbatim" because the value has already gone through
  the writer once.
- Add method outputVerbatim to the writer interface and classes
- Add method checkCloseLength to the writer interface and classes. The
  method is necessary because when the closeTag function gets called, it
  creates a leading greater than sign and a newline to the output, which
  should be skipped for the DataFor field.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

mayx commented:
@afishbeck Hi Tony please code review when you have a chance...

richardkchapman commented:
@jakesmith Can you take a look too

richardkchapman commented on common/thorhelper/thorxmlwrite.hpp:
```diff
@@ -125,6 +127,9 @@ class thorhelper_decl CommonXmlWriter : implements IXmlWriterExt, public CInterf
     }
 
+    virtual size32_t checkCloseLength(unsigned len, const char* data);
+    virtual void outputVerbatim(unsigned len, const char* data) {out.append(len, data);}
```
minor: normally add spaces after { and before }

afishbeck replied:
I think you want a closeTag(); before the append.

jakesmith commented on esp/esdllib/esdl_transformer2.cpp:
```diff
@@ -1111,8 +1111,12 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
+                                    StringBuffer databuf(ctx.writer->str()+len);
                                     ctx.writer->rewindTo(location);
+                                    unsigned closeLen = ctx.writer->checkCloseLength(databuf.length(), databuf);
+                                    if(closeLen > databuf.length())
```
trivial/formatting: space between if(

jakesmith commented on esp/esdllib/esdl_transformer2.cpp:
```diff
@@ -1111,8 +1111,12 @@ void Esdl2Struct::process(Esdl2TransformerContext &ctx, const char *out_name, Es
                                 if (ctx.writer->length() > len)
```
Given dataForProcessed has been flagged since len set, is it possible for this condition to be false?

jakesmith commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -70,6 +70,19 @@ void CommonXmlWriter::closeTag()
 
+size32_t CommonXmlWriter::checkCloseLength(unsigned len, const char* data)
+{
+    int closeLen = 0;
```
should be size32_t

jakesmith commented:
@mayx - am a bit dubious whether right approach, or logic or placement of checkCloseLength().

Instead of allowing datafor to be written and then working hard to un-write it.
Where it sets the flag (dataForProcessed) could you instead setup a dataForWriter and write to a dataForWriter instead directly, so never written to the main writer?

afishbeck commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -70,6 +70,19 @@ void CommonXmlWriter::closeTag()
     flush(false);
 }
 
+size32_t CommonXmlWriter::checkCloseLength(unsigned len, const char* data)
```
@mayx I'd like to try to think of a better abstraction for what we're really trying to do here.  May be tricky.

afishbeck commented:
@jakesmith I don't think we want to change the underlying mechanism in this time frame.

Basically what is happening is that we've already built the content, but we want to cut and paste it into another location.  If I remember correctly the rule is that this pasted content will always end up at the bottom so that this process can always work as a stream.

@mayx I think the abstraction works better if we combine the rewind with the "cut" functionality, and the internal parser can deal with the close bracket if it needs to.

So
StringBuffer databuf(ctx.writer->str()+len);
ctx.writer->rewindTo(location);
unsigned closeLen = ctx.writer->checkCloseLength(databuf.length(), databuf);
if(closeLen > databuf.length())
  closeLen = databuf.length();
local.setDataFor(chd.queryDataFor()->queryName(), databuf.str() + closeLen);

becomes

StringBuffer databuf;
ctx.writer->cutFrom(location, databuf);
local.setDataFor(chd.queryDataFor()->queryName(), databuf);

afishbeck commented on common/thorhelper/thorxmlwrite.hpp:
```diff
@@ -41,6 +41,8 @@ interface IXmlWriterExt : extends IXmlWriter
     virtual void rewindTo(IInterface *location) = 0;
     virtual void outputNumericString(const char *field, const char *fieldname) = 0;
+    virtual size32_t checkCloseLength(unsigned len, const char* data) = 0;
+    virtual void outputVerbatim(unsigned len, const char* data) = 0;
```
I think outputInline() or something similar would be a better name.  This is basically a format neutral version of the existing "outputInlineXml()" function.

jakesmith commented:
@afishbeck >I don't think we want to change the underlying mechanism in this time frame.

Fair enough, your cutFrom suggestion would make it a lot cleaner.

jakesmith commented:
@mayx - can you make that change?

mayx commented:
Change checked in.

mayx commented:
(1) fixed format issues (2) added functions cutFrom and appendInline, removed checkCloseLength and appendVerbatiom. (3) Jake the if condition you mentioned is necessary in the case when the item tag exists but is empty. @jakesmith @afishbeck Please review as soon as you can, we'd like to make the change into the coming build.

afishbeck commented on common/thorhelper/thorxmlwrite.hpp:
```diff
@@ -71,6 +73,7 @@ class thorhelper_decl CommonXmlWriter : implements IXmlWriterExt, public CInterf
     void outputEndNested(const char *fieldname, bool doIndent);
 
     virtual void outputInlineXml(const char *text){closeTag(); out.append(text); flush(false);} //for appending raw xml content
+    virtual void outputInline(const char* text) { closeTag(); out.append(text); flush(false); }
```
could just call outputInlineXml(text).

mayx replied:
I debated about that. OK, changing now...

jakesmith commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -70,6 +70,30 @@ void CommonXmlWriter::closeTag()
+    if (flusher)
+        throwUnexpected();
+
+    CommonXmlPosition *position = dynamic_cast<CommonXmlPosition *>(location);
```
Perhaps too big a change now, but feels lke saveLocation / position should be a proper interface with methods to access pos, not dynamic casting up like this.

afishbeck replied:
Maybe.  I had purposely made it opaque, not wanting callers to make any assumptions about how they could make use of the location from outside.  Other implementations could save very different bits of information.  But open for discussion... for future changes.

mayx committed:
HPCC-17855 DynamicESDL support for get_data_from

- Output DataFor inline (as is) because the value has already gone through
  the writer once.
- Add method outputInline to the writer interface and classes
- Add method cutFrom to the writer interface and classes to retract data
  and rewind position at the same time.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

jakesmith commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -70,6 +70,30 @@ void CommonXmlWriter::closeTag()
+            startInd += 1;
+        if(!position->nestLimit && startInd < out.length() && out.charAt(startInd) == '\n')
+            startInd += 1;
+        if(startInd < out.length())
```
trivial/formatting: space between if and '(' in a few places.

mayx commented:
@afishbeck  Updated.

jakesmith commented on common/thorhelper/thorxmlwrite.cpp:
```diff
@@ -596,6 +620,24 @@ void CommonJsonWriter::outputEndDataset(const char *dsname)
+        needDelimiter = position->needDelimiter;
+        indent = position->indent;
+        nestLimit = position->nestLimit;
+    }
```
Is the above code identical to that in CommonXmlWriter?
Should be commoned up iso, maybe via a base class?

jakesmith commented:
@mayx  - don't think anything above need prevent it from merging, but worth revisiting.
